### PR TITLE
Simulated Annealing

### PR DESCRIPTION
**Pull request by darioizzo**

### CODE REVIEW

File: include/pagmo/algorithms/uda_template.hpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
+/* Copyright 2017 PaGMO development team
```
Comment (bluescarni) [inline include/pagmo/algorithms/uda_template.hpp]: Is this file supposed to be included in the repo?

File: include/pagmo/algorithms/simulated_annealing.hpp [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+     * Constructs simulated_annealing
+     *
+     * @param Ts starting temperature
+     * @param Tf final temperature
+     * @param n_T_adj number of temperature adjustments in the annealing schedule
+     * @param n_range_adj number of adjustments of the search range performed at a constant temperature
+     * @param bin_size number of mutations that are used to compute the acceptance rate
+     * @param start_range starting range for mutating the decision vector
+     * @param seed seed used by the internal random number generator
+     */
```
Comment (bluescarni) [inline include/pagmo/algorithms/simulated_annealing.hpp]: Could you add the throw specifications, as done in the pygmo exposition?

File: include/pagmo/algorithms/simulated_annealing.hpp [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+                                                   + std::to_string(start_range) + " was detected.");
+        }
+    }
+
+    /// Algorithm evolve method (juice implementation of the algorithm)
+    /**
+     * @param pop population to be evolved
+     * @return evolved population
+     * @throws std::invalid_argument if the problem is multi-objective or constrained
+     */
```
Comment (bluescarni) [inline include/pagmo/algorithms/simulated_annealing.hpp]: It throws also if the problem is stochastic or pop size < 1, right?

File: pygmo/core.cpp [hunk 4]
```diff
@@ -729,6 +730,17 @@ BOOST_PYTHON_MODULE(core)
...
+    auto simulated_annealing_ = pygmo::expose_algorithm<simulated_annealing>(
+        "simulated_annealing", pygmo::simulated_annealing_docstring().c_str());
+    simulated_annealing_.def(bp::init<double, double, unsigned, unsigned, unsigned, double>(
+        (bp::arg("Ts") = 10., bp::arg("Tf") = 0.1, bp::arg("n_T_adj") = 10u, bp::arg("n_range_adj") = 1u,
+         bp::arg("bin_size") = 20u, bp::arg("start_range") = 1.)));
+    simulated_annealing_.def(bp::init<double, double, unsigned, unsigned, unsigned, double>(
+        (bp::arg("Ts") = 10., bp::arg("Tf") = 0.1, bp::arg("n_T_adj") = 10u, bp::arg("n_range_adj") = 10u,
+         bp::arg("bin_size") = 10u, bp::arg("start_range") = 1., bp::arg("seed"))));
+    pygmo::expose_algo_log(simulated_annealing_, pygmo::simulated_annealing_get_log_docstring().c_str());
+    simulated_annealing_.def("get_seed", &sea::get_seed);
```
Comment (bluescarni) [inline pygmo/core.cpp]: Shouldn't this be ``&simulated_annealing::get_seed()``?

File: tests/simulated_annealing.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+using namespace pagmo;
+
+BOOST_AUTO_TEST_CASE(simulated_annealing_construction)
+{
+    BOOST_CHECK_NO_THROW(simulated_annealing{});
+    simulated_annealing user_algo{10, 0.1, 10u, 10u, 10u, 1., 23u};
+    BOOST_CHECK(user_algo.get_verbosity() == 0u);
+    BOOST_CHECK(user_algo.get_seed() == 23u);
+    // BOOST_CHECK((user_algo.get_log() == cmaes::log_type{}));
```
Comment (bluescarni) [inline tests/simulated_annealing.cpp]: Should this test be restored?

File: include/pagmo/algorithms/simulated_annealing.hpp [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+        m_log.clear();
+
+        std::uniform_real_distribution<double> drng(0., 1.); // to generate a number in [0, 1)
+
+        // Starting point is the best individual
+        auto best_idx = pop.best_idx();
+        const auto &x0 = pop.get_x()[best_idx];
+        const auto &fit0 = pop.get_f()[best_idx];
+        // Determines the coefficient to decrease the temperature
+        const double Tcoeff = std::pow(m_Tf / m_Ts, 1.0 / static_cast<double>(m_n_T_adj));
```
Comment (darioizzo) [inline include/pagmo/algorithms/simulated_annealing.hpp]: Fixed by requesting m_Ts and m_n_T_adj be strictly positive

File: include/pagmo/algorithms/uda_template.hpp (hunk 1 above)
Comment (darioizzo) [inline include/pagmo/algorithms/uda_template.hpp]: probably not

**Pull request merged**